Control/lib/services/environment/EnvironmentCache.service.js

```diff
@@ -39,6 +40,40 @@ class EnvironmentCacheService {
     this._listenToEventsAndBroadcast();
   }
 
+  /**
+   * Setter for updating existing environments in cache
+   * @param {EnvironmentInfo[]} environments - the environments to be updated
+   */
+  set environments(environments) {
+    environments.forEach((environment) => {
+      const { id } = environment;
+      let cachedEnvironment = {};
+      if (this._environments.has(id)) {
+        cachedEnvironment = this._environments.get(id);
+        const { events = [] } = cachedEnvironment;
+        Object.assign(cachedEnvironment, environment);
+        cachedEnvironment.events = events;  
+      } else {
+        cachedEnvironment = { ...environment };
+        if (!cachedEnvironment.events) {
+          cachedEnvironment.events = [];
+        }
+      }
+
+      this._environments.set(id, cachedEnvironment);
```

Comment:
I think it's safer to avoid modifying the existing object directly

```suggestion
      let updatedEnvironment = {};
const existing = this._environments.get(id);

if (existing) {
  updatedEnvironment = {
    ...existing,
    ...environment,
    events: existing.events || [],
  };
} else {
  updatedEnvironment = {
    ...environment,
    events: environment.events || [],
  };
}

this._environments.set(id, updatedEnvironment);

```

Reply:
Thank you for the feedback! I have updated the object.assign to create a new target